baremaps-server/src/main/java/com/baremaps/server/CorsFilter.java

```diff
@@ -1,28 +1,63 @@
+/*
+ * Code inspired by resteasy CorsFilter: https://github.com/resteasy/Resteasy/blob/22984e96a2f40dc556ba7b1613422febd94f86d2/resteasy-core/src/main/java/org/jboss/resteasy/plugins/interceptors/CorsFilter.java
+ *
+ Copyright (c) Resteasy Maintainers. All rights reserved.
+ Licensed under the ASL license.
+ */
+
 package com.baremaps.server;
 
-import java.io.IOException;
 import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
 import javax.ws.rs.container.ContainerResponseContext;
 import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.core.Response;
 import javax.ws.rs.ext.Provider;
 
 @Provider
-public class CorsFilter implements ContainerResponseFilter {
+public class CorsFilter implements ContainerRequestFilter, ContainerResponseFilter {
+
+  public static class CorsHeaders {
+    public static final String ACCESS_CONTROL_ALLOW_ORIGIN = "Access-Control-Allow-Origin";
+    public static final String ACCESS_CONTROL_ALLOW_CREDENTIALS = "Access-Control-Allow-Credentials";
+    public static final String ACCESS_CONTROL_ALLOW_METHODS = "Access-Control-Allow-Methods";
+    public static final String ACCESS_CONTROL_ALLOW_HEADERS = "Access-Control-Allow-Headers";
+    public static final String ORIGIN = "Origin";
+    public static final String VARY = "Vary";
+  }
+
+  @Override
+  public void filter(ContainerRequestContext requestContext) {
+    String origin = requestContext.getHeaderString(CorsHeaders.ORIGIN);
+    if (origin == null) {
+      return;
+    }
+    if (requestContext.getMethod().equalsIgnoreCase("OPTIONS")) {
+      preflight(origin, requestContext);
+    }
+  }
 
   @Override
-  public void filter(ContainerRequestContext containerRequestContext, ContainerResponseContext containerResponseContext)
-      throws IOException {
-    containerResponseContext.getHeaders().add(
-        "Access-Control-Allow-Origin",
-        "*");
-    containerResponseContext.getHeaders().add(
-        "Access-Control-Allow-Credentials",
-        "true");
-    containerResponseContext.getHeaders().add(
-        "Access-Control-Allow-Headers",
-        "origin, content-type, accept, authorization");
-    containerResponseContext.getHeaders().add(
-        "Access-Control-Allow-Methods",
-        "GET, POST, PUT, DELETE, OPTIONS, HEAD");
+  public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
+    String origin = requestContext.getHeaderString(CorsHeaders.ORIGIN);
+    if (origin == null || requestContext.getMethod().equalsIgnoreCase("OPTIONS") || requestContext.getProperty("cors.failure") != null) {
+      return;
+    }
+    responseContext.getHeaders().putSingle(CorsHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
```

Comment:
We may move CorsHeaders' constants in the CorsFilter class to make these lines shorter (java's verbosity...), but it is also fine for me as it is.

```java
var headers = responseContext.getHeaders();
headers.putSingle(ACCESS_CONTROL_ALLOW_ORIGIN, origin);
headers.putSingle(ACCESS_CONTROL_ALLOW_METHODS, "GET, POST, PUT, DELETE, OPTIONS, HEAD");
...

Reply:
I didn't know I could do that !